// @soos-io/api-client/dist/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnFailure = exports.OutputFormat = exports.LogLevel = exports.SeverityEnum = exports.ScmType = exports.ScmResultsFormat = exports.ScanType = exports.ScanStatus = exports.ManifestStatus = exports.PackageManagerType = exports.IntegrationType = exports.IntegrationName = exports.ContributingDeveloperSource = void 0;
var ContributingDeveloperSource;
(function (ContributingDeveloperSource) {
    ContributingDeveloperSource["Unknown"] = "Unknown";
    ContributingDeveloperSource["GitHubWebhook"] = "GitHubWebhook";
    ContributingDeveloperSource["EnvironmentVariable"] = "EnvironmentVariable";
    ContributingDeveloperSource["OperatingSystem"] = "OperatingSystem";
})(ContributingDeveloperSource || (exports.ContributingDeveloperSource = ContributingDeveloperSource = {}));
var IntegrationName;
(function (IntegrationName) {
    IntegrationName["AzureDevOps"] = "AzureDevOps";
    IntegrationName["AWSCodeBuild"] = "AWSCodeBuild";
    IntegrationName["Bamboo"] = "Bamboo";
    IntegrationName["BitBucket"] = "BitBucket";
    IntegrationName["CircleCI"] = "CircleCI";
    IntegrationName["CodeShip"] = "CodeShip";
    IntegrationName["GithubActions"] = "GitHub";
    IntegrationName["GitLab"] = "GitLab";
    IntegrationName["Jenkins"] = "Jenkins";
    IntegrationName["SoosCsa"] = "SoosCsa";
    IntegrationName["SoosDast"] = "SoosDast";
    IntegrationName["SoosSast"] = "SoosSast";
    IntegrationName["SoosSca"] = "SoosSca";
    IntegrationName["SoosSbom"] = "SoosSbom";
    IntegrationName["TeamCity"] = "TeamCity";
    IntegrationName["TravisCI"] = "TravisCI";
    IntegrationName["VisualStudio"] = "VisualStudio";
    IntegrationName["VisualStudioCode"] = "VisualStudioCode";
})(IntegrationName || (exports.IntegrationName = IntegrationName = {}));
var IntegrationType;
(function (IntegrationType) {
    IntegrationType["None"] = "None";
    IntegrationType["IDE"] = "IDE";
    IntegrationType["Script"] = "Script";
    IntegrationType["Webhook"] = "Webhook";
    IntegrationType["Plugin"] = "Plugin";
    IntegrationType["AppRepo"] = "AppRepo";
    IntegrationType["AppUpload"] = "AppUpload";
})(IntegrationType || (exports.IntegrationType = IntegrationType = {}));
var PackageManagerType;
(function (PackageManagerType) {
    PackageManagerType["Unknown"] = "Unknown";
    PackageManagerType["Alpine"] = "Alpine";
    PackageManagerType["CFamily"] = "CFamily";
    PackageManagerType["Dart"] = "Dart";
    PackageManagerType["Debian"] = "Debian";
    PackageManagerType["Docker"] = "Docker";
    PackageManagerType["Erlang"] = "Erlang";
    PackageManagerType["Fedora"] = "Fedora";
    PackageManagerType["Go"] = "Go";
    PackageManagerType["Homebrew"] = "Homebrew";
    PackageManagerType["Java"] = "Java";
    PackageManagerType["NPM"] = "NPM";
    PackageManagerType["NuGet"] = "NuGet";
    PackageManagerType["Php"] = "Php";
    PackageManagerType["Python"] = "Python";
    PackageManagerType["Ruby"] = "Ruby";
    PackageManagerType["Rust"] = "Rust";
    PackageManagerType["Swift"] = "Swift";
    PackageManagerType["Wolfi"] = "Wolfi";
})(PackageManagerType || (exports.PackageManagerType = PackageManagerType = {}));
var ManifestStatus;
(function (ManifestStatus) {
    ManifestStatus["Unknown"] = "Unknown";
    ManifestStatus["Valid"] = "Valid";
    ManifestStatus["OnlyDevDependencies"] = "OnlyDevDependencies";
    ManifestStatus["OnlyLockFiles"] = "OnlyLockFiles";
    ManifestStatus["OnlyNonLockFiles"] = "OnlyNonLockFiles";
    ManifestStatus["NoPackages"] = "NoPackages";
    ManifestStatus["UnknownManifestType"] = "UnknownManifestType";
    ManifestStatus["UnsupportedManifestVersion"] = "UnsupportedManifestVersion";
    ManifestStatus["ParsingError"] = "ParsingError";
    ManifestStatus["Empty"] = "Empty";
})(ManifestStatus || (exports.ManifestStatus = ManifestStatus = {}));
var ScanStatus;
(function (ScanStatus) {
    ScanStatus["Unknown"] = "Unknown";
    ScanStatus["Queued"] = "Queued";
    ScanStatus["Manifest"] = "Manifest";
    ScanStatus["LocatingDependencies"] = "LocatingDependencies";
    ScanStatus["LoadingPackageDetails"] = "LoadingPackageDetails";
    ScanStatus["LocatingVulnerabilities"] = "LocatingVulnerabilities";
    ScanStatus["RunningGovernancePolicies"] = "RunningGovernancePolicies";
    ScanStatus["Finished"] = "Finished";
    ScanStatus["FailedWithIssues"] = "FailedWithIssues";
    ScanStatus["Incomplete"] = "Incomplete";
    ScanStatus["Error"] = "Error";
})(ScanStatus || (exports.ScanStatus = ScanStatus = {}));
var ScanType;
(function (ScanType) {
    ScanType["CSA"] = "Csa";
    ScanType["DAST"] = "Dast";
    ScanType["SAST"] = "Sast";
    ScanType["SBOM"] = "Sbom";
    ScanType["SCA"] = "Sca";
})(ScanType || (exports.ScanType = ScanType = {}));
var ScmResultsFormat;
(function (ScmResultsFormat) {
    ScmResultsFormat["JSON"] = "JSON";
    ScmResultsFormat["TXT"] = "TXT";
})(ScmResultsFormat || (exports.ScmResultsFormat = ScmResultsFormat = {}));
var ScmType;
(function (ScmType) {
    ScmType["GitHub"] = "GitHub";
    ScmType["BitbucketCloud"] = "BitbucketCloud";
})(ScmType || (exports.ScmType = ScmType = {}));
var SeverityEnum;
(function (SeverityEnum) {
    SeverityEnum["Unknown"] = "Unknown";
    SeverityEnum["None"] = "None";
    SeverityEnum["Info"] = "Info";
    SeverityEnum["Low"] = "Low";
    SeverityEnum["Medium"] = "Medium";
    SeverityEnum["High"] = "High";
    SeverityEnum["Critical"] = "Critical";
})(SeverityEnum || (exports.SeverityEnum = SeverityEnum = {}));
var LogLevel;
(function (LogLevel) {
    LogLevel["DEBUG"] = "DEBUG";
    LogLevel["INFO"] = "INFO";
    LogLevel["WARN"] = "WARN";
    LogLevel["FAIL"] = "FAIL";
    LogLevel["ERROR"] = "ERROR";
})(LogLevel || (exports.LogLevel = LogLevel = {}));
var OutputFormat;
(function (OutputFormat) {
    OutputFormat["SARIF"] = "SARIF";
})(OutputFormat || (exports.OutputFormat = OutputFormat = {}));
var OnFailure;
(function (OnFailure) {
    OnFailure["Continue"] = "continue_on_failure";
    OnFailure["Fail"] = "fail_the_build";
})(OnFailure || (exports.OnFailure = OnFailure = {}));
